### PR TITLE
Prometheus: Move count_values from function to aggregation

### PR DESCRIPTION
**Pull request by aocenas**
count_values was defined as function but it should have been an aggregation.
Contains also some changes to the aggregation factory function and moved code that is already shared with loki to /shared dir. 

Also fixes bug, where aggregations change to their `_by` variant every time user blured other param input.

### CODE REVIEW

File: public/app/plugins/datasource/prometheus/querybuilder/aggregations.ts [hunk 1]
```diff
@@ -36,162 +39,6 @@ export function getAggregationOperations(): QueryBuilderOperationDef[] {
   ];
 }
 
-/**
```
Comment (aocenas) [inline public/app/plugins/datasource/prometheus/querybuilder/aggregations.ts:39]: All this code was moved to shared/operationsUtils as it is already used in loki

File: public/app/plugins/datasource/prometheus/querybuilder/shared/operationUtils.ts [hunk 2]
```diff
@@ -144,3 +148,169 @@ export function getRangeVectorParamDef(withRateInterval = false): QueryBuilderOp
...
+      addOperationHandler: defaultAddOperationHandler,
+      hideFromList: true,
+      ...overrides,
+    },
+  ];
+
+  return operations;
+}
+
+export function createAggregationOperationWithParam(
```
Comment (aocenas) [inline public/app/plugins/datasource/prometheus/querybuilder/shared/operationUtils.ts:229]: The main change is this function. Before this code was inside `createAggregationOperation` and just special cased for topK/bottomK aggregations.

File: public/app/plugins/datasource/prometheus/querybuilder/shared/operationUtils.ts [hunk 3]
```diff
@@ -144,3 +148,169 @@ export function getRangeVectorParamDef(withRateInterval = false): QueryBuilderOp
...
+  name: string,
+  paramsDef: { params: QueryBuilderOperationParamDef[]; defaultParams: QueryBuilderOperationParamValue[] },
+  overrides: Partial<QueryBuilderOperationDef> = {}
+): QueryBuilderOperationDef[] {
+  const operations = createAggregationOperation(name, overrides);
+  operations[0].params.unshift(...paramsDef.params);
+  operations[1].params.unshift(...paramsDef.params);
+  operations[2].params.unshift(...paramsDef.params);
```
Comment (aocenas) [inline public/app/plugins/datasource/prometheus/querybuilder/shared/operationUtils.ts:237]: Before we weren't adding params to the `without` variant for some reason so this adds it also there.

File: public/app/plugins/datasource/prometheus/querybuilder/shared/operationUtils.ts [hunk 4]
```diff
@@ -144,3 +148,169 @@ export function getRangeVectorParamDef(withRateInterval = false): QueryBuilderOp
...
+    const params = model.params.slice(0, -1);
+    const restParams = model.params.slice(1);
+    return `${aggregation} by(${restParams.join(', ')}) (${params.map(mapType).join(', ')}, ${innerExpr})`;
+  };
+}
+
+/**
+ * This function will transform operations without labels to their plan aggregation operation
+ */
+function getLastLabelRemovedHandler(changeToOperartionId: string) {
```
Comment (ivanahuckova) [inline public/app/plugins/datasource/prometheus/querybuilder/shared/operationUtils.ts:294]:
```suggestion
function getLastLabelRemovedHandler(changeToOperationId: string) {
```

Comment (ivanahuckova) [reply public/app/plugins/datasource/prometheus/querybuilder/shared/operationUtils.ts:294]: replace all `changeToOperartionId` -> `changeToOperartionId `

File: public/app/plugins/datasource/prometheus/querybuilder/shared/operationUtils.ts [hunk 5]
```diff
@@ -144,3 +148,169 @@ export function getRangeVectorParamDef(withRateInterval = false): QueryBuilderOp
...
+        ...op,
+        id: changeToOperartionId,
+      };
+    }
+
+    return op;
+  };
+}
+
+function getOnLabelAdddedHandler(changeToOperartionId: string) {
```
Comment (ivanahuckova) [inline public/app/plugins/datasource/prometheus/querybuilder/shared/operationUtils.ts:309]:
```suggestion
function getOnLabelAdddedHandler(changeToOperationId: string) {
```

replace all `changeToOperartionId` -> `changeToOperartionId `

Comment (ivanahuckova) [inline public/app/plugins/datasource/prometheus/querybuilder/shared/operationUtils.ts:309]:
```suggestion
function getOnLabelAddedHandler(changeToOperartionId: string) {
```

Comment (ivanahuckova) [reply public/app/plugins/datasource/prometheus/querybuilder/shared/operationUtils.ts:309]: replace all 'getOnLabelAdddedHandler` -> `getOnLabelAddedHandler`

**Review by ivanahuckova — CHANGES_REQUESTED** [review]
Left some spell-check fixes/feedback below. Overall looks good, but it breaks Loki's topk/bottomk operations because it still probably uses `createAggregationOperation` instead of new `createAggregationOperationWithParam`. 
image

File: public/app/plugins/datasource/prometheus/querybuilder/shared/operationUtils.ts [hunk 6]
```diff
@@ -144,3 +148,175 @@ export function getRangeVectorParamDef(withRateInterval = false): QueryBuilderOp
...
+    return op;
+  };
+}
+
+function getOnLabelAddedHandler(changeToOperationId: string) {
+  return function onParamChanged(index: number, op: QueryBuilderOperation, def: QueryBuilderOperationDef) {
+    // Check if we actually have the label param. As it's optional the aggregation can have one less, which is the
+    // case of just simple aggregation without label. When user adds the label it now has the same number of params
+    // as it's definition, and now we can change it to it's `_by` variant.
+    if (op.params.length === def.params.length) {
```
Comment (aocenas) [inline public/app/plugins/datasource/prometheus/querybuilder/shared/operationUtils.ts:314]: Semi-related fix. As this wasn't checked before any time this was called (like on input onBlur) the type of the operation changed.

**Review by ivanahuckova — CHANGES_REQUESTED** [review]
There is still an issue with Loki query builder. On main, when you select `TopK` aggregation, it adds also range function `rate`. After your change, this is not a a case anymore. 

Main: 

https://user-images.githubusercontent.com/30407135/161990851-be791db1-9445-4723-8cc0-6bc6b63b62e3.mov

Your branch: 

https://user-images.githubusercontent.com/30407135/161990883-2e7013e2-e0ff-44b5-8041-eaf35273fd30.mov

**Review by ivanahuckova — APPROVED** [review]
Nice, LGTM. 🙌 We probably want to backport this to 8.5.0.

**Pull request merged**